src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusScaleMonitor.cs

```diff
@@ -0,0 +1,337 @@
+            }
+
+            return CreateTriggerMetrics(metrics, isUnauthorized, _isListeningOnDeadLetterQueue);
+        }
+
+        private async Task<ServiceBusMetricsData> GetQueueMetricsAsync()
+        {
+            ServiceBusMetricsData metrics = new ServiceBusMetricsData(null, 0, 0, 0);
+            QueueRuntimeInfo queueRuntimeInfo;
+            QueueDescription queueDescription;
+
+            // Peek the first message in the queue without removing it from the queue
+            metrics.Message = await _receiver.Value.PeekAsync();
+
+            queueRuntimeInfo = await _managementClient.Value.GetQueueRuntimeInfoAsync(_entityPath);
+            metrics.ActiveMessageCount = queueRuntimeInfo.MessageCountDetails.ActiveMessageCount;
+            metrics.DeadLetterCount = queueRuntimeInfo.MessageCountDetails.DeadLetterMessageCount;
+
+            // If partitioning is turned on, then Service Bus automatically partitions queues into 16 partitions
+            // See more information here: https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-partitioning#standard
+            queueDescription = await _managementClient.Value.GetQueueAsync(_entityPath);
+            metrics.PartitionCount = queueDescription.EnablePartitioning ? 16 : 0;
+
+            return metrics;
+        }
+
+        private async Task<ServiceBusMetricsData> GetTopicMetricsAsync()
+        {
+            ServiceBusMetricsData metrics = new ServiceBusMetricsData(null, 0, 0, 0);
+            TopicDescription topicDescription;
+            SubscriptionRuntimeInfo subscriptionRuntimeInfo;
+            string topicPath, subscriptionPath;
+
+            // Peek the first message in the subscription without removing it from the queue
+            metrics.Message = await _receiver.Value.PeekAsync();
+
+            ServiceBusEntityPathHelper.ParseTopicAndSubscription(_entityPath, out topicPath, out subscriptionPath);
+
+            subscriptionRuntimeInfo = await _managementClient.Value.GetSubscriptionRuntimeInfoAsync(topicPath, subscriptionPath);
+            metrics.ActiveMessageCount = subscriptionRuntimeInfo.MessageCountDetails.ActiveMessageCount;
+            metrics.DeadLetterCount = subscriptionRuntimeInfo.MessageCountDetails.DeadLetterMessageCount;
+
+            // If partitioning is turned on, then Service Bus automatically partitions queues into 16 partitions
+            // See more information here: https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-partitioning#standard
+            topicDescription = await _managementClient.Value.GetTopicAsync(topicPath);
+            metrics.PartitionCount = topicDescription.EnablePartitioning ? 16 : 0;
+
+            return metrics;
+        }
+
+        public static ServiceBusTriggerMetrics CreateTriggerMetrics(ServiceBusMetricsData metrics, bool isUnauthorized, bool isListeningOnDeadLetterQueue)

```

Comment:
I'd just get rid of this ServiceBusMetricsData class, and instead pass the bits of info needed into this helper. Then GetQueueMetrics/GetTopicMetrics just call this helper and return a ScaleMetrics instance directly, w/o the need for an intermediate type.

Reply:
Yeah, that would be better. The only problem is that `CreateTriggerMetrics` needs the `isUnauthorized` flag. But we only get that flag in a catch block in `GetMetricsAsync`. I could move just the `isUnauthorized` catch to `GetQueueMetrics`/`GetTopicMetrics`, so we'd have the original try/catch split across those three functions, would that be cleaner than the intermediate type?

Reply:
I'm trying to understand why you even need the isUnauthorized bool. You're only using it to set the totalNewEventCount to zero, right? But if an exception is thrown trying to get the count, it'll be zero anyways since it was never set (exception from management client call)?

Reply:
So I re-examined the scale controller code and you're right, the isUnauthorized bool isn't necessary with the way we have it set up. I also caught a few bugs in my implementation testing the no manage claim scenario. Without a manage claim, we can peek a message but can't get info about the queue/runtime (active, dead, and partition counts). So we can peek the message in GetMetricsAsync and provide an alternate path to CreateTriggerMetrics when an Unauthroized exception is thrown in GetQueueMetrics/GetTopicMetrics. No managed claim should have MessageCount = 1 and QueueTime = some TimeSpan, which this should accomplish